src/main/java/io/cdap/plugin/gcp/bigquery/action/BigQueryExecute.java

```diff
@@ -0,0 +1,209 @@
+      // errors, not just the latest one.
+      throw new RuntimeException(queryJob.getStatus().getExecutionErrors().toString());
+    }
+
+    TableResult queryResults = queryJob.getQueryResults();
+    long rows = queryResults.getTotalRows();
+
+    context.getMetrics().gauge(context.getStageName() + RECORDS_OUT, rows);
+  }
+
+  @Override
+  public void configurePipeline(PipelineConfigurer configurer) throws IllegalArgumentException {
+    config.validate();
+  }
+
+  /**
+   * Config for the plugin.
+   */
+  public final class Config extends GCPConfig {
+    @Description("Use Legacy SQL.")
+    @Macro
+    private String legacy;
+
+    @Description("BigQuery SQL to execute.")
+    @Macro
+    private String sql;
+
+    @Description("Mode to execute the query in. The value must be 'batch' or 'interactive'. " +
+                   "A batch query is executed as soon as possible and count towards the concurrent rate limit " +
+                   "and the daily rate limit. An interactive query is queued and started as soon as idle resources " +
+                   "are available, usually within a few minutes. If the query hasn't started within 3 hours, " +
+                   "its priority is changed to 'interactive'")
+    @Macro
+    private String mode;
+
+    @Description("Use the cache when executing the query.")
+    @Macro
+    private String useCache;
+
+    @Description("Location of the job. Must match the location of the dataset specified in the query. Defaults to 'US'")
+    @Macro
+    private String location;
+
+    @Description("The dataset to store the query results in. If not specified, the results will not be stored.")
+    @Macro
+    @Nullable
+    private String dataset;
+
+    @Description("The table to store the query results in. If not specified, the results will not be stored.")
+    @Macro
+    @Nullable
+    private String table;
+
+    public boolean isLegacySQL() {
+      return legacy.equalsIgnoreCase("true");
+    }
+
+    public boolean shouldUseCache() {
+      return useCache.equalsIgnoreCase("true");
+    }
+
+    public String getLocation() {
+      return location;
+    }
+
+    public String getSql() {
+      return sql;
+    }
+
+    public QueryJobConfiguration.Priority getMode() {
+      return QueryJobConfiguration.Priority.valueOf(mode.toUpperCase());
+    }
+
+    @Nullable
+    public String getDataset() {
+      return dataset;
+    }
+
+    @Nullable
+    public String getTable() {
+      return table;
+    }
+
+    public void validate() {
+      // check the mode is valid
+      if (!containsMacro("mode")) {
+        getMode();
+      }
+
+      if (!containsMacro("sql") && (sql == null || sql.isEmpty())) {
+        throw new IllegalArgumentException("SQL not specified. Please specify a SQL to execute");
+      }
+
+      // validates that either they are null together or not null together
+      if ((dataset == null && table != null) || (table == null && dataset != null)) {
```

Comment:
Option string config properties should use null and empty checks.`Strings.isNullOrEmpty`

Reply:
this should also check for `containsMacro`.